numpy/core/src/umath/loops.c.src

```diff
@@ -1642,26 +1642,26 @@ pairwise_sum_@TYPE@(@dtype@ *a, npy_uintp n, npy_intp stride)
          * 8 times unroll reduces blocksize to 16 and allows vectorization with
          * avx without changing summation ordering
          */
-        r[0] = @trf@(a[0 * stride]);
-        r[1] = @trf@(a[1 * stride]);
-        r[2] = @trf@(a[2 * stride]);
-        r[3] = @trf@(a[3 * stride]);
-        r[4] = @trf@(a[4 * stride]);
-        r[5] = @trf@(a[5 * stride]);
-        r[6] = @trf@(a[6 * stride]);
-        r[7] = @trf@(a[7 * stride]);
+        r[0] = @trf@(*((@dtype@ *)(a + 0 * stride)));
+        r[1] = @trf@(*((@dtype@ *)(a + 1 * stride)));
+        r[2] = @trf@(*((@dtype@ *)(a + 2 * stride)));
+        r[3] = @trf@(*((@dtype@ *)(a + 3 * stride)));
+        r[4] = @trf@(*((@dtype@ *)(a + 4 * stride)));
+        r[5] = @trf@(*((@dtype@ *)(a + 5 * stride)));
+        r[6] = @trf@(*((@dtype@ *)(a + 6 * stride)));
+        r[7] = @trf@(*((@dtype@ *)(a + 7 * stride)));
 
         for (i = 8; i < n - (n % 8); i += 8) {
             /* small blocksizes seems to mess with hardware prefetch */
-            NPY_PREFETCH(&a[(i + 512 / sizeof(a[0])) * stride], 0, 3);
-            r[0] += @trf@(a[(i + 0) * stride]);
-            r[1] += @trf@(a[(i + 1) * stride]);
-            r[2] += @trf@(a[(i + 2) * stride]);
-            r[3] += @trf@(a[(i + 3) * stride]);
-            r[4] += @trf@(a[(i + 4) * stride]);
-            r[5] += @trf@(a[(i + 5) * stride]);
-            r[6] += @trf@(a[(i + 6) * stride]);
-            r[7] += @trf@(a[(i + 7) * stride]);
+            NPY_PREFETCH(a + (i + 512) * stride, 0, 3);
```

Comment:
After the original, I think this should be `NPY_PREFETCH(a + (i + 512 / sizeof(@dtype@)) * stride, 0, 3);` . I don't know that it makes much difference in practice.

Reply:
Yeah I think you're right. This confused me since when comparing the original and new versions, not only are we changing `a`'s type, but we are also redefining the `stride` variable to reflect the stride in bytes rather than the stride in `itemsize`s. 

The original code referred to memory position (in bytes) `a + (i + 512 / sizeof(a[0])) * (stride/sizeof(a[0]))  * sizeof(a[0])` where that last `sizeof` term accounts for C array indexing behavior for a pointer of dtype type, and where `stride` is the array stride in bytes. 

In the new code, we use a `char*` pointer, so should absorb the last sizeof into the expression, which should give  `a + (i + 512 / sizeof(@dtype@)) * stride` as you said.

Reply:
Honestly, I think this is the sort of thing a good compiler should handle as part of loop optimization. It's all a bit fuzzy, with optimum depending on latency and other execution factors. Getting it wrong can even have a negative impact.

Reply:
I think my [earlier comment](https://github.com/numpy/numpy/pull/9887#discussion_r145600166) was perhaps correct after all then